src/sqlfluff/core/parser/segments/base.py

```diff
@@ -1199,7 +1199,7 @@ def apply_fixes(
 
             before = []
             after = []
-            if not r.can_start_end_non_code:
+            if not r.can_start_end_non_code and r.parse_grammar:
```

Comment:
Can you explain why this is necessary and how the lack of it caused the bug? Bit confused.

Reply:
Agree. The intent of this code is to avoid parse errors. If you look at the other place where we check this flag (in the parse() function about 200 lines above), you'll see that it only does this check if there's a parse grammar. So I can't answer the deeper "why", but the simple answer is to match that behavior, only changing things if not doing so would raise an error later there.

Reply:
I'll see about adding a comment and/or changing the two areas of code to be more clearly "aligned".

Reply:
Comment added, code tweaked _slightly_ to align a bit more with `parse()`.